src/Extensions/Hosting/Certificates/CertificateReader.cs

```diff
@@ -0,0 +1,117 @@
+
+		public CertificateReader(ICertificateStore certificateStore, ILogger<CertificateReader> logger)
+		{
+			m_certificateStore = certificateStore;
+			m_logger = logger;
+			m_certificatesCache = new ConcurrentDictionary<StoreName, CertificateInformation[]>();
+		}
+
+		public X509Certificate2? GetCertificateByThumbprint(string thumbprint, bool refreshCache, StoreName storeName)
+		{
+			Validation.ThrowIfNullOrWhiteSpace(thumbprint, nameof(thumbprint));
+
+			X509Certificate2? matchingCertificate = null;
+			foreach (CertificateInformation info in LoadCertificates(storeName, refreshCache))
+			{
+				if (string.Equals(info.Certificate.Thumbprint, thumbprint, StringComparison.OrdinalIgnoreCase))
+				{
+					matchingCertificate = info.Certificate;
+					break;
+				}
+			}
+
+			if (matchingCertificate == null)
+			{
+				m_logger.LogError(Tag.Create(), "Could not find certificate with {0} thumbprint.", thumbprint);
+			}
+			else
+			{
+				m_logger.LogDebug(Tag.Create(), "Found certificate with {0} thumbprint.", thumbprint);
+			}
+
+			return matchingCertificate;
+		}
+
+		public IEnumerable<X509Certificate2> GetCertificatesByCommonName(string commonName, bool refreshCache, StoreName storeName)
+		{
+			Validation.ThrowIfNullOrWhiteSpace(commonName, nameof(commonName));
+
+			bool certFound = false;
+			foreach (CertificateInformation info in LoadCertificates(storeName, refreshCache))
+			{
+				if (string.Equals(info.CommonName, commonName, StringComparison.Ordinal))
+				{
+					certFound = true;
+					yield return info.Certificate;
+				}
+			}
+
+			if (certFound)
+			{
+				m_logger.LogDebug(Tag.Create(), "Found at least one certificate with {0} as common name.", commonName);
+			}
+			else
+			{
+				m_logger.LogError(Tag.Create(), "Could not find certificate with {0} as common name.", commonName);
+			}
+		}
+
+		public X509Certificate2? GetCertificateByCommonName(string commonName, bool refreshCache, StoreName storeName) =>
+			PickBestCertificate(GetCertificatesByCommonName(commonName, refreshCache, storeName));
+
+		private CertificateInformation[] LoadCertificates(StoreName storeName, bool refreshCache)
+		{
+			if (refreshCache || !m_certificatesCache.TryGetValue(storeName, out CertificateInformation[] certificates))
+			{
+				m_logger.LogInformation(Tag.Create(), "Updating certificates cache for store '{0}'.", storeName);
+
+				certificates = m_certificateStore.GetAllCertificates(storeName, StoreLocation.LocalMachine).ToArray();
+
+				m_logger.LogInformation(Tag.Create(), "Inserting '{0}' certificates to certificates cache for store '{1}'.", certificates.Length, storeName);
+
+				m_certificatesCache[storeName] = certificates;
+			}
+
+			return certificates ?? Array.Empty<CertificateInformation>();
+		}
+
+		private static X509Certificate2? PickBestCertificate(IEnumerable<X509Certificate2> certificates)
+		{
+			DateTime now = DateTime.Now;
```

Comment:
> DateTime.Now; [](start = 17, length = 14)

For testability, it is better to inject the time

Reply:
I've considered it, but in this case, does not any useful functionality since you can build unit tests around DataTime.Now, but creating certificates with `NotBefore = DateTime.Now.AddDays(-1)` to get required state